async.go

```diff
@@ -90,3 +105,52 @@ func panicHandler(ctx context.Context, errCh chan error) {
 func getContextWithTaskName(ctx context.Context, taskName string) context.Context {
 	return loggers.AddToLogContext(ctx, "taskname", taskName)
 }
+
+func GoWithWait(ctx context.Context,
+	taskNames []string,
+	tasks []func(context.Context) error,
+	timeout time.Duration,
+	logger log.Logger) error {
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer func() {
+		if ctx.Err() == nil {
+			cancel()
+		}
+	}()
+
+	errs := make(chan error, len(tasks))
+
+	wg := new(sync.WaitGroup)
+	waitCh := make(chan struct{})
+	wg.Add(len(tasks))
+
+	go func() {
+		for i, task := range tasks {
+			taskToRun := task
+			taskName := taskNames[i]
+			go func() {
+				defer wg.Done()
+				taskErr := execute(ctx, taskToRun, taskName, logger)
+				if taskErr != nil && ctx.Err() == nil {
+					errs <- taskErr
+					cancel()
+				}
+			}()
+		}
+
+		wg.Wait()
+		close(waitCh)
+	}()
+
+	select {
+	case <-waitCh: // successfully completed all tasks
+		close(waitCh)
+		return nil
+	case err := <-errs:
+		close(errs)
```

Comment:
shouldn't we close waitCh too ?

Reply:
guaranteed when error out ? or panic ?

Reply:
it will be closed after when wg.Done() executed for sure but in case of error we will return immediately without waiting for wg.Done() to complete

Reply:
done

Reply:
we should close the channels if its open.

Reply:
if close the waitCh immediately after getting error then go routine will panic after wg.Done() since goroutine is still executing in background it will try to close waitCh

Reply:
so we are leaving that channel open ?

Reply:
added defer close(errs) to close error channel